[FLINK-28588] Enhance REST API for Speculative Execution

pltbkd opened the pull request:
<!--
*Thank you very much for contributing to Apache Flink - we are happy that you want to help us improve Flink. To help the community review your contribution in the best possible way, please go through the checklist below, which will get the contribution into a shape in which it can be best reviewed.*

*Please understand that we do not do this to make contributions to Flink a hassle. In order to uphold a high standard of quality for code contributions, while at the same time managing a large number of contributions, we need contributors to prepare the contributions well, and give reviewers enough contextual information for the review. Please also understand that contributions that do not follow this guide will take longer to review and thus typically be picked up with lower priority by the community.*

## Contribution Checklist

  - Make sure that the pull request corresponds to a [JIRA issue](https://issues.apache.org/jira/projects/FLINK/issues). Exceptions are made for typos in JavaDoc or documentation files, which need no JIRA issue.
  
  - Name the pull request in the form "[FLINK-XXXX] [component] Title of the pull request", where *FLINK-XXXX* should be replaced by the actual issue number. Skip *component* if you are unsure about which is the best component.
  Typo fixes that have no associated JIRA issue should be named following this pattern: `[hotfix] [docs] Fix typo in event time introduction` or `[hotfix] [javadocs] Expand JavaDoc for PuncuatedWatermarkGenerator`.

  - Fill out the template below to describe the changes contributed by the pull request. That will give reviewers the context they need to do the review.
  
  - Make sure that the change passes the automated tests, i.e., `mvn clean verify` passes. You can set up Azure Pipelines CI to do that following [this guide](https://cwiki.apache.org/confluence/display/FLINK/Azure+Pipelines#AzurePipelines-Tutorial:SettingupAzurePipelinesforaforkoftheFlinkrepository).

  - Each pull request should address only one issue, not mix up code from multiple issues.
  
  - Each commit in the pull request has a meaningful commit message (including the JIRA id)

  - Once all items of the checklist are addressed, remove the above text and this checklist, leaving only the filled out template below.


**(The sections below can be removed for hotfixes of typos)**
-->

## What is the purpose of the change

As a follow-up step of FLIP-168 and FLIP-224, the Flink Web UI needs to be enhanced to display the related information if the speculative execution mechanism is enabled. This PR is introducing the changes of REST APIs.

Note: Commits except the last one is for Blocklist Mechanism, and will be removed once its related pull requests are merged.

## Brief change log

  - A "other-concurrent-attempts" field is added to SubtaskExecutionAttemptDetailsInfo and SubtaskBackPressureInfo, to hold the information of other concurrent attempts other than the fastest one.
  - Count of blocked task managers and slots is added to the ResourceOverview.
  - Blocked flag is added to the TaskManagerInfo, indicating if the task manager is blocked.
  - MetricStore is modified to hold the metrics of more than one current execution attempt.

## Verifying this change

This change added tests and can be verified by running all unit tests.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? yes
  - If yes, how is the feature documented? The feature is documented in another pr.

zhuzhurk commented on flink-runtime/src/main/java/org/apache/flink/runtime/webmonitor/WebMonitorEndpoint.java:
```diff
@@ -495,6 +496,10 @@ protected List<Tuple2<RestHandlerSpecification, ChannelInboundHandler>> initiali
         final SubtaskMetricsHandler subtaskMetricsHandler =
                 new SubtaskMetricsHandler(leaderRetriever, timeout, responseHeaders, metricFetcher);
 
+        final SubtaskAttemptMetricsHandler subtaskAttemptMetricsHandler =
```
I think this REST API is not in the scope of this FLIP. Therefore, I think we should remove it.

zhuzhurk commented on flink-runtime/src/test/java/org/apache/flink/runtime/executiongraph/ArchivedExecutionVertexWithSpeculativeExecutionTest.java:
```diff
@@ -0,0 +1,436 @@
+
+    private static void compareExecutionVertex(
+            AccessExecutionVertex runtimeVertex, AccessExecutionVertex archivedVertex) {
+        assertEquals(
```
JUnit assertions should be avoided. See https://flink.apache.org/contributing/code-style-and-quality-common.html#testing.

zhuzhurk commented on flink-runtime/src/test/java/org/apache/flink/runtime/messages/webmonitor/JobDetailsTest.java:
```diff
@@ -103,4 +105,33 @@ public void testJobDetailsCompatibleUnmarshalling() throws IOException {
+
+        final JobDetails unmarshalled = objectMapper.treeToValue(marshalled, JobDetails.class);
+
+        assertEquals(expected, unmarshalled);
```
Flink is migrating to JUnit 5 so that all the new tests should use `JUnit5` and `assertJ`.
You can migrate the existing tests to JUnit5 in a  hotfix commit before making functional changes.

zhuzhurk reviewed (commented):
Thanks for opening this PR! @pltbkd 
I haven't fully reviewed the changes, but have a few comments that I think should be addressed earlier.

pltbkd commented on flink-runtime/src/main/java/org/apache/flink/runtime/rest/messages/taskmanager/TaskManagerInfo.java:
```diff
@@ -113,7 +123,8 @@ public TaskManagerInfo(
             @JsonProperty(FIELD_NAME_HARDWARE) HardwareDescription hardwareDescription,
-            @JsonProperty(FIELD_NAME_MEMORY) TaskExecutorMemoryConfiguration memoryConfiguration) {
+            @JsonProperty(FIELD_NAME_MEMORY) TaskExecutorMemoryConfiguration memoryConfiguration,
+            @JsonProperty(FIELD_NAME_BLOCKED) @Nullable Boolean blocked) {
```
With @JsonInclude(Include.NON_DEFAULT), the json may not include the blocked field when it's false. We add the annotation since we don't want to change the REST API reponse for non speculative execution users.
While when deserializing, jackson will try to apply a null to the field when the field is absent, and it will cause exception if the field is a boolean.

pltbkd commented:
@zhuzhurk 
Thanks a lot for the reviewing! 
I have split the pr into more meaning commits, and most of your suggestions have been taken, except for the jackson one, which I have replied with the reason.
Would you please take another look at the PR?

pltbkd commented on flink-runtime/src/test/java/org/apache/flink/runtime/rest/handler/legacy/messages/ClusterOverviewWithVersionTest.java:
```diff
@@ -31,6 +31,6 @@ protected Class<ClusterOverviewWithVersion> getTestResponseClass() {
     @Override
     protected ClusterOverviewWithVersion getTestResponseInstance() {
-        return new ClusterOverviewWithVersion(1, 3, 3, 7, 4, 2, 0, "version", "commit");
+        return new ClusterOverviewWithVersion(2, 6, 3, 1, 3, 7, 4, 2, 0, "version", "commit");
```
I think we have to modify because the blocked tms and available tms are expected to be not 0 for the testing, so the totalTaskManager has to be more than 1.

pltbkd commented on flink-runtime/src/main/java/org/apache/flink/runtime/rest/messages/taskmanager/TaskManagerDetailsInfo.java:
```diff
@@ -64,6 +66,7 @@ public TaskManagerDetailsInfo(
             @JsonProperty(FIELD_NAME_AVAILABLE_RESOURCE) ResourceProfileInfo freeResource,
             @JsonProperty(FIELD_NAME_HARDWARE) HardwareDescription hardwareDescription,
             @JsonProperty(FIELD_NAME_MEMORY) TaskExecutorMemoryConfiguration memoryConfiguration,
+            @JsonProperty(FIELD_NAME_BLOCKED) @Nullable Boolean blocked,
```
It's the same case as above

zhuzhurk commented on flink-runtime/src/test/java/org/apache/flink/runtime/executiongraph/ArchivedExecutionGraphTest.java:
```diff
@@ -42,43 +41,34 @@
+import static org.assertj.core.api.Assertions.assertThat;
 
 /** Tests for the {@link ArchivedExecutionGraph}. */
 public class ArchivedExecutionGraphTest extends TestLogger {
```
`public` is no longer needed for JUnit5 tests.
`extends TestLogger` is no longer needed as well.

zhuzhurk commented on flink-runtime/src/test/java/org/apache/flink/runtime/executiongraph/ArchivedExecutionGraphTest.java:
```diff
@@ -42,43 +41,34 @@
 
-    @BeforeClass
+    @BeforeAll
     public static void setupExecutionGraph() throws Exception {
```
`Before*`/`After*` methods are not needed to be public in JUnit5.

zhuzhurk commented on flink-runtime/src/test/java/org/apache/flink/runtime/executiongraph/ArchivedExecutionGraphTest.java:
```diff
@@ -159,19 +149,11 @@ public void testSerialization() throws IOException, ClassNotFoundException {
-
-        assertThat(suspendedExecutionGraph.getState(), is(JobStatus.SUSPENDED));
-        assertThat(suspendedExecutionGraph.getFailureInfo(), notNullValue());
+    public static void assertContainsCheckpointSettings(ArchivedExecutionGraph archivedGraph) {
```
`Test` methods are not needed to be public in JUnit5.

zhuzhurk commented on flink-runtime/src/test/java/org/apache/flink/runtime/executiongraph/ArchivedExecutionGraphTest.java:
```diff
@@ -159,19 +149,11 @@ public void testSerialization() throws IOException, ClassNotFoundException {
     }
 
-    @Test
-    public void testCreateFromInitializingJobForSuspendedJob() {
```
Moving around these methods makes hard to compare the diff of the changes.

pltbkd replied:
It must be a mistake. Sorry for that and I'll move them back.

zhuzhurk commented on flink-runtime/src/test/java/org/apache/flink/runtime/messages/webmonitor/JobDetailsTest.java:
```diff
@@ -79,7 +79,7 @@ public void testJobDetailsMarshalling() throws JsonProcessingException {
         final JobDetails unmarshalled = objectMapper.treeToValue(marshalled, JobDetails.class);
 
-        assertEquals(expected, unmarshalled);
+        assertThat(expected).isEqualTo(unmarshalled);
```
-> assertThat(unmarshalled).isEqualTo(expected);

zhuzhurk commented on flink-runtime/src/test/java/org/apache/flink/runtime/messages/webmonitor/JobDetailsTest.java:
```diff
@@ -101,6 +101,6 @@ public void testJobDetailsCompatibleUnmarshalling() throws IOException {
                 objectMapper.readValue(COMPATIBLE_JOB_DETAILS, JobDetails.class);
 
-        assertEquals(expected, unmarshalled);
+        assertThat(expected).isEqualTo(unmarshalled);
```
-> assertThat(unmarshalled).isEqualTo(expected);

zhuzhurk commented on flink-runtime/src/main/java/org/apache/flink/runtime/rest/handler/job/JobVertexTaskManagersHandler.java:
```diff
@@ -182,30 +188,32 @@ private static JobVertexTaskManagersInfo createJobVertexTaskManagersInfo(
-                                current.getAccumulateBackPressuredTime(),
-                                current.getAccumulateIdleTime(),
-                                current.getAccumulateBusyTime()));
+                for (AccessExecution attempt : vertex.getCurrentExecutions()) {
```
Looks to me this is not right because the attempts of an execution vertex may locate in different task managers.

pltbkd replied:
You are right. While when I was trying to do a further refactor of this handler, I found that maybe we should use only the representing execution to calculate the aggregated task manager status, or the state can be confusing since if one attempt is CANCELED (even it's a speculative one), the state of the task manager is CANCELED. I don't think this is the expected behavior. I'd revert the changes and add a comment to describe why we choose to use the representing attempt only.

zhuzhurk replied:
I think it can help to display all the attempts on that TM. And then we should also include them into the aggregated IO metrics of the TM. Regarding the status shown on that tab, I think it can be an aggregation of all the **representative** attempts on that task manger, so that it can be FINISHED if the job successfully finishes.

pltbkd replied:
I think it's better. I'll refactor the method in this way.

zhuzhurk commented on flink-runtime/src/main/java/org/apache/flink/runtime/rest/handler/job/SubtaskExecutionAttemptAccumulatorsHandler.java:
```diff
@@ -99,25 +99,27 @@ public Collection<ArchivedJson> archiveJsonWithPath(AccessExecutionGraph graph)
-
-                archive.add(new ArchivedJson(curAttemptPath, curAttemptJson));
+                for (AccessExecution attempt : subtask.getCurrentExecutions()) {
+                    if (attempt != null) {
```
In which case `attempt` can be null?

pltbkd replied:
It refers to the HistoricalExecutions part, but indeed it's redundant since the attempt should never be null. I'll remove the check.

zhuzhurk commented on flink-runtime/src/main/java/org/apache/flink/runtime/rest/handler/job/SubtaskExecutionAttemptDetailsHandler.java:
```diff
@@ -111,36 +111,44 @@ public Collection<ArchivedJson> archiveJsonWithPath(AccessExecutionGraph graph)
-
-                archive.add(new ArchivedJson(curAttemptPath, curAttemptJson));
+                for (AccessExecution attempt : subtask.getCurrentExecutions()) {
+                    if (attempt != null) {
```
In which case attempt can be null?

zhuzhurk commented on flink-runtime/src/main/java/org/apache/flink/runtime/rest/messages/job/SubtaskExecutionAttemptDetailsInfo.java:
```diff
@@ -100,6 +106,36 @@ public class SubtaskExecutionAttemptDetailsInfo implements ResponseBody {
+    @Nullable
+    private final List<SubtaskExecutionAttemptDetailsInfo> otherConcurrentAttempts;
+
+    public SubtaskExecutionAttemptDetailsInfo(
```
Is this for test only? If so, I would suggest to remove it and refactor the tests instead.

zhuzhurk commented on flink-runtime/src/main/java/org/apache/flink/runtime/webmonitor/threadinfo/JobVertexThreadInfoTracker.java:
```diff
@@ -257,18 +258,19 @@ private Map<TaskManagerLocation, ImmutableSet<ExecutionAttemptID>> groupExecutio
-            }
-            Set<ExecutionAttemptID> groupedAttemptIds =
-                    executionAttemptsByLocation.getOrDefault(tmLocation, new HashSet<>());
+            for (AccessExecution execution : executionVertex.getCurrentExecutions()) {
```
To double confirm, would any problem occur if one subtask has multiple attempts?

pltbkd replied:
The ExecutionAttemptID stands for a specific attempt of a specific execution. It seems the further processing doesn't care if there are ExecutionAttemptIDs belonging to the same execution. It processes each ExecutionAttemptID independently.

zhuzhurk replied:
I'm mainly not sure whether the web UI can correctly handle this.

zhuzhurk commented on flink-runtime/src/main/java/org/apache/flink/runtime/rest/handler/job/JobVertexBackPressureHandler.java:
```diff
@@ -100,26 +106,73 @@ private JobVertexBackPressureInfo createJobVertexBackPressureInfo(
+                                otherConcurrentAttempts));
+            }
         }
         result.sort(Comparator.comparingInt(SubtaskBackPressureInfo::getSubtask));
```
Is this a stable sort?

pltbkd replied:
Yes it is since there's only one element in the result for each subtask. Other attempts are in the otherConcurrentAttempts.

zhuzhurk commented on flink-runtime/src/main/java/org/apache/flink/runtime/rest/handler/job/JobVertexBackPressureHandler.java:
```diff
@@ -100,26 +106,73 @@ private JobVertexBackPressureInfo createJobVertexBackPressureInfo(
 
+    private SubtaskBackPressureInfo createSubtaskAttemptBackpressureInfo(
+            int subtaskIndex,
+            Integer attemptNumber,
```
Better to annotate it and otherConcurrentAttempts as Nullable.

zhuzhurk commented on flink-runtime/src/main/java/org/apache/flink/runtime/rest/handler/job/JobVertexBackPressureHandler.java:
```diff
@@ -100,26 +106,73 @@ private JobVertexBackPressureInfo createJobVertexBackPressureInfo(
+                                subtaskIndex, null, subtaskMetricStore, null));
+            } else {
+                int currentAttempt =
+                        currentExecutionAttempts == null
```
In which case currentExecutionAttempts can be null?

pltbkd replied:
currentExecutionAttempts is acquired from the JobDetails. In fact the currentExecutionAttempts of a subtask only exists when there's more than one current attempt. So it's normal that currentExecutionAttempts is null.

zhuzhurk commented on flink-runtime/src/main/java/org/apache/flink/runtime/rest/handler/job/JobVertexBackPressureHandler.java:
```diff
@@ -100,26 +106,73 @@ private JobVertexBackPressureInfo createJobVertexBackPressureInfo(
+                        currentExecutionAttempts == null
+                                ? -1
+                                : currentExecutionAttempts.getOrDefault(subtaskIndex, -1);
+                if (!allAttemptsMetricStores.containsKey(currentAttempt)) {
```
Is there any case that `currentAttempt != -1` but it is not contained in `allAttemptsMetricStores`?

pltbkd replied:
Since JobDetails and MetricStore are acquired independently, maybe the attempt exists when acquiring JobDetails while it's already failed when acquiring the MetricStore.

zhuzhurk replied:
So this means it can happen that the displayed representative attempt in the WebUI is not the true representative attempt? e.g. A CANCELED attempt is displayed as representative attempt while there is another RUNNING/FINISHED attempt.

pltbkd replied:
I don't think so. The issue only exists while fetching the metrics of a subtask or a task, that the acquired metrics may be not from the latest representative attempt. The archived execution graph contains which attempt is the representative one, and the metrics are acquired with attempt number. So neither the state nor the metrics for each attempt will be wrong.

pltbkd commented on flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/ArchivedExecutionVertex.java:
```diff
@@ -40,15 +44,27 @@ public class ArchivedExecutionVertex implements AccessExecutionVertex, Serializa
+        currentExecutions = new ArrayList<>(vertex.getCurrentExecutions().size());
+        currentExecution = vertexCurrentExecution.archive();
+        currentExecutions.add(currentExecution);
+        for (Execution execution : vertex.getCurrentExecutions()) {
```
The representing one has already been added a few lines above. Since Execution.archive creates a new ArchivedExecution, to ensure the currentExecutions contains the currentExecution, it's added there and excluded here.

pltbkd commented on flink-runtime/src/main/java/org/apache/flink/runtime/messages/webmonitor/ClusterOverview.java:
```diff
@@ -44,11 +50,22 @@ public class ClusterOverview extends JobsOverview {
             @JsonProperty(FIELD_NAME_TASKMANAGERS) int numTaskManagersConnected,
             @JsonProperty(FIELD_NAME_SLOTS_TOTAL) int numSlotsTotal,
             @JsonProperty(FIELD_NAME_SLOTS_AVAILABLE) int numSlotsAvailable,
+            @JsonProperty(FIELD_NAME_TASKMANAGERS_BLOCKED) @Nullable Integer numTaskManagersBlocked,
```
The constructor is used as the json creator by jackson. If the json doesn't contain the field, jackson will use null as the field value, which will cause a NPE. We may meet the case if HistoryServer archive the graph with an older version of flink while read it with the new version.

pltbkd replied:
The constructor is used as the json creator by jackson. If the json doesn't contain the field, jackson will use null as the field value, which will cause a NPE. We may meet the case if HistoryServer archive the graph with an older version of flink while read it with the new version.

pltbkd commented on flink-runtime/src/main/java/org/apache/flink/runtime/rest/handler/legacy/metrics/MetricStore.java:
```diff
@@ -156,6 +179,27 @@ public synchronized ComponentMetricStore getSubtaskMetricStore(
+            return null;
+        }
+        SubtaskMetricStore subtask = task.getSubtaskMetricStore(subtaskIndex);
+        if (attemptId < 0) {
```
A negative attempt number is considered valid before when there were constructors using -1 as the default value, and was used to acquire the subtask level metrics instead of the attempt level one. Now that we have removed those constructors, a negative attempt  number is indeed invalid. I'll cleanup the relating codes.

pltbkd commented on flink-runtime/src/main/java/org/apache/flink/runtime/rest/handler/legacy/metrics/MetricStore.java:
```diff
@@ -177,7 +221,9 @@ public void add(MetricDump metric) {
-            ComponentMetricStore subtask;
+            SubtaskMetricStore subtask;
+            ComponentMetricStore attempt;
+            boolean isCurrentAttempt = true;
```
isRepresentativeAttempt here is a good name to explain the meaning. But I would limit the "representative" naming here. CurrentExecutionAttempts in the JobDetails is collected via Execution#getCurrentExecutionAttempt, and  passed to other places. Maybe we can modify the naming when we formally introduce the concept of the representative attempt for an Execution.

pltbkd commented on flink-runtime/src/main/java/org/apache/flink/runtime/rest/handler/legacy/metrics/MetricStore.java:
```diff
@@ -54,6 +57,14 @@ public class MetricStore {
+     * vertex. When a metric of an execution attempt is added, the metric can also be added to the
+     * SubtaskMetricStore when it is of the representing execution.
+     */
+    private final Map<String, Map<String, Map<Integer, Integer>>> currentExecutionAttempts =
```
I'd prefer to remain the naming. See the reply above.
Or maybe we may only modify the map here in the MetricStore, but do not modify that in other classes?

pltbkd commented on flink-runtime/src/main/java/org/apache/flink/runtime/rest/handler/job/JobVertexTaskManagersHandler.java:
```diff
@@ -131,18 +132,23 @@ private static JobVertexTaskManagersInfo createJobVertexTaskManagersInfo(
+                taskManagerId2Host.put(taskmanagerId, taskManagerHost);
+                List<AccessExecutionVertex> vertices =
+                        taskManagerVertices.computeIfAbsent(
+                                taskmanagerId, ignored -> new ArrayList<>(4));
```
Though I agree, I'd prefer not to change the existing code.

zhuzhurk replied:
It should be avoided according to the code style guide.
https://flink.apache.org/contributing/code-style-and-quality-java.html#collections

pltbkd replied:
ok, I'll remove it

gaoyunhaii commented on flink-runtime/src/main/java/org/apache/flink/runtime/rest/messages/job/SubtaskExecutionAttemptDetailsInfo.java:
```diff
@@ -261,6 +261,53 @@ public static SubtaskExecutionAttemptDetailsInfo create(
+                otherConcurrentAttempts);
+    }
+
+    public List<SubtaskExecutionAttemptDetailsInfo> getOtherConcurrentAttempts() {
```
This method seems not called?

pltbkd replied:
Quite a few getters in this class are not called. I'm not sure if we should add the getter or not. I'll move this method next to other getters first. Please let me know if it should actually be removed.

zhuzhurk commented on flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/ArchivedExecutionVertex.java:
```diff
@@ -79,6 +96,11 @@ public ArchivedExecution getCurrentExecutionAttempt() {
 
+    @Override
+    public Collection<AccessExecution> getCurrentExecutions() {
+        return Collections.unmodifiableCollection(currentExecutions);
```
I'm not too concerned about this because it's relatively lightweight and has been widely adopted in runtime code, as long as it will not be invoked too frequently (e.g. O(N) per RPC).

pltbkd commented on flink-runtime/src/main/java/org/apache/flink/runtime/rest/handler/legacy/metrics/MetricStore.java:
```diff
@@ -391,4 +476,36 @@ private static TaskMetricStore unmodifiable(TaskMetricStore source) {
+            this.attempts = checkNotNull(attempts);
+        }
+
+        private static SubtaskMetricStore unmodifiable(SubtaskMetricStore source) {
```
It's a bug not using the method in getSubtaskMetricStore. Now it's fixed so the method is used.

pltbkd commented on flink-runtime/src/main/java/org/apache/flink/runtime/rest/handler/job/JobVertexDetailsHandler.java:
```diff
@@ -120,9 +120,24 @@ private static JobVertexDetailsInfo createJobVertexDetailsInfo(
             subtasks.add(
                     SubtaskExecutionAttemptDetailsInfo.create(
-                            execution, metricFetcher, jobID, jobVertexID));
+                            execution, metricFetcher, jobID, jobVertexID, otherConcurrentAttempts));
```
I'd prefer not to change the existing code that declaring the variable, as long as the naming is not really confusing here.

pltbkd commented:
Hi @zhuzhurk and @gaoyunhaii ,
Thanks for the reviewing again!
A new fixup commit has been pushed. It's resolving most of the comments. I'll separate and squash the fixes into original commits once the review is done.

pltbkd commented on flink-runtime/src/main/java/org/apache/flink/runtime/rest/handler/job/JobExceptionsHandler.java:
```diff
@@ -127,22 +128,24 @@ private static JobExceptionsInfoWithHistory createJobExceptionsInfo(
+                    taskExceptionList.add(
+                            new JobExceptionsInfo.ExecutionExceptionInfo(
+                                    failure.get().getExceptionAsString(),
+                                    task.getTaskNameWithSubtaskIndex(),
```
I'm not sure if we should modify it here in this pr. The change is beneficial while seems not relating to the speculative execution. Maybe it's better to do as a separate issue?

zhuzhurk commented on flink-runtime/src/main/java/org/apache/flink/runtime/rest/handler/job/JobVertexTaskManagersHandler.java:
```diff
@@ -175,45 +180,48 @@ private static JobVertexTaskManagersInfo createJobVertexTaskManagersInfo(
+                executionsPerState[state.ordinal()]++;
+                if (representativeExecutions.contains(execution)) {
+                    tasksPerState[state.ordinal()]++;
+                    representativeAttemptsCount += 1;
```
nit: representativeAttemptsCount++;

zhuzhurk reviewed (approved):
LGTM. Thanks for addressing all the comments! @pltbkd

pltbkd commented:
Thanks you for all the comments!

I've squashed all fixups and rebased on to the latest master. The PR should be ready for merging once the CI is passed.

gaoyunhaii reviewed (approved):
Thanks @pltbkd for the fix! Also LGTM from my side~
